### PR TITLE
⚠️ Default Kubelet cgroupDriver to systemd for Kubernetes >= 1.21

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it:**
back port of https://github.com/kubernetes-sigs/cluster-api/pull/4236

For Kubernetes >= 1.21, image builder is going to setup containerd for using systemd cgroup.
KCP should take care of adapting the Kubelet-Config map to be consistent with this change.

**Which issue(s) this PR fixes:**
Rif #4099

/area control-plane
/milestone v0.3.x

### CODE REVIEW

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 1]
```diff
@@ -42,16 +43,20 @@ import (
...
 	kubeProxyKey              = "kube-proxy"
 	kubeadmConfigKey          = "kubeadm-config"
+	kubeletConfigKey          = "kubelet"
+	cgroupDriverKey           = "cgroupDriver"
 	labelNodeRoleControlPlane = "node-role.kubernetes.io/master"
 )
 
 var (
-	ErrControlPlaneMinNodes = errors.New("cluster has fewer than 2 control plane nodes; removing an etcd member is not supported")
+	minVerKubeletSystemdDriver = semver.MustParse("1.21.0")
```
Comment (neolit123) [inline controlplane/kubeadm/internal/workload_cluster.go:58]: this version can be inline in ErrControlPlaneMinNodes()

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 2]
```diff
@@ -174,6 +179,42 @@ func (w *Workload) UpdateKubeletConfigMap(ctx context.Context, version semver.Ve
...
+		kubeletConfig, err := yamlToUnstructured([]byte(data))
+		if err != nil {
+			return errors.Wrapf(err, "unable to decode kubelet ConfigMap's %q content to Unstructured object", kubeletConfigKey)
+		}
+		cgroupDriver, _, err := unstructured.NestedString(kubeletConfig.UnstructuredContent(), cgroupDriverKey)
+		if err != nil {
+			return errors.Wrapf(err, "unable to extract %q from Kubelet ConfigMap's %q", cgroupDriverKey, cm.Name)
+		}
+
+		// If the value is not already explicitly set by the user, change according to kubeadm/image builder new requirements.
```
Comment (neolit123) [inline controlplane/kubeadm/internal/workload_cluster.go:203]:
```suggestion
		// If the value is not already explicitly set by the user, change according to the kubeadm/image builder new requirements.
```

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 3]
```diff
@@ -174,6 +179,42 @@ func (w *Workload) UpdateKubeletConfigMap(ctx context.Context, version semver.Ve
...
+		if cgroupDriver == "" {
+			cgroupDriver = "systemd"
+
+			if err := unstructured.SetNestedField(kubeletConfig.UnstructuredContent(), cgroupDriver, cgroupDriverKey); err != nil {
+				return errors.Wrapf(err, "unable to update %q on Kubelet ConfigMap's %q", cgroupDriverKey, cm.Name)
```
Comment (neolit123) [inline controlplane/kubeadm/internal/workload_cluster.go:208]:
```suggestion
				return errors.Wrapf(err, "unable to update %q in the kubelet ConfigMap %q", cgroupDriverKey, cm.Name)
```

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 4]
```diff
@@ -174,6 +179,42 @@ func (w *Workload) UpdateKubeletConfigMap(ctx context.Context, version semver.Ve
...
+			}
+			updated, err := yaml.Marshal(kubeletConfig)
+			if err != nil {
+				return errors.Wrapf(err, "unable to encode Kubelet ConfigMap's %q to YAML", cm.Name)
```
Comment (neolit123) [inline controlplane/kubeadm/internal/workload_cluster.go:212]:
```suggestion
				return errors.Wrapf(err, "unable to encode the kubelet ConfigMap %q to YAML", cm.Name)
```

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 5]
```diff
@@ -174,6 +179,42 @@ func (w *Workload) UpdateKubeletConfigMap(ctx context.Context, version semver.Ve
...
+		if !ok {
+			return errors.Errorf("unable to find %q key in %s", kubeletConfigKey, cm.Name)
+		}
+		kubeletConfig, err := yamlToUnstructured([]byte(data))
+		if err != nil {
+			return errors.Wrapf(err, "unable to decode kubelet ConfigMap's %q content to Unstructured object", kubeletConfigKey)
+		}
+		cgroupDriver, _, err := unstructured.NestedString(kubeletConfig.UnstructuredContent(), cgroupDriverKey)
+		if err != nil {
+			return errors.Wrapf(err, "unable to extract %q from Kubelet ConfigMap's %q", cgroupDriverKey, cm.Name)
```
Comment (neolit123) [inline controlplane/kubeadm/internal/workload_cluster.go:200]:
```suggestion
			return errors.Wrapf(err, "unable to extract %q from the kubelet ConfigMap %q", cgroupDriverKey, cm.Name)
```

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 6]
```diff
@@ -174,6 +179,42 @@ func (w *Workload) UpdateKubeletConfigMap(ctx context.Context, version semver.Ve
...
+	// because only new nodes using v1.21 images will pick up the change during
+	// kubeadm join.
+	if version.GE(minVerKubeletSystemdDriver) {
+		data, ok := cm.Data[kubeletConfigKey]
+		if !ok {
+			return errors.Errorf("unable to find %q key in %s", kubeletConfigKey, cm.Name)
+		}
+		kubeletConfig, err := yamlToUnstructured([]byte(data))
+		if err != nil {
+			return errors.Wrapf(err, "unable to decode kubelet ConfigMap's %q content to Unstructured object", kubeletConfigKey)
```
Comment (neolit123) [inline controlplane/kubeadm/internal/workload_cluster.go:196]:
```suggestion
			return errors.Wrapf(err, "unable to decode the kubelet ConfigMap %q content to Unstructured object", kubeletConfigKey)
```

**Comment by neolit123** [comment]
technically, the CAPI change is not needed as long as `kubeadm init` from 1.21 defaults the `KubeletConfiguration` that it uploads in the cluster to have the `systemd` driver. if kubeadm is doing that it will match the image-builder driver and i don't see a need for CAPI to do that too?

File: controlplane/kubeadm/internal/workload_cluster.go (hunk 1 above)
Comment (fabriziopandini) [reply controlplane/kubeadm/internal/workload_cluster.go:58]: this is a backport, let's keep thing consistent given this is a minor

**Comment by fabriziopandini** [comment]
> maybe this code path is done during upgrade, where an old config map has to be migrated to match the image builder driver?

this is the case

WRT to the minor fix, given this is a backport I prefer keep things consistent; eventually we can fix in main and backport those nits well

**Review by vincepri — APPROVED** [review]
/approve
/assign @neolit123 
for lgtm

**Pull request merged**